### ISSUE TITLE
Make LineUp more configurable

### ISSUE THREAD

**Issue by puehringer** [lineupjs/lineupjs#230]
* Release number or git hash: v4.0.0-alpha.12
 * Web browser version and OS: 

**This is a summary ticket which should be split into smaller tickets after specification. Specification is still WIP.**

### Observed and expected behavior
While many options are already available (such as `rowHeight`), many other options (such as `defaultColumnWidth`) are still missing. 

Additionally, columns, dialogs, renderers, ... should be more configurable (including setters). 
For columns, it should be possible to specify which menu options are available (currently solved via annotations but not configurable --> maybe change to dynamic `addMenuItem`). Also their text, icon, ... should come from a configuration. 
For dialogs, it should be possible to specify defaults. I.e. for string filtering, it is not possible to auto-check the regex checkbox (or hide it). One could override the dialog with a custom dialog, but then code-duplication is an issue. Another approach could be hooks called pre/post the rendering, and pre/post the saving (to modify it before applying), but `showRegexCheckbox` in a `dialogs.<stringFilterIdentifier>` configuration object would be an easier and easily implementable approach. 

In short: LineUp should be much more configurable (labels, defaults, ...)

**Comment by thinkh** [comment]
In the following some notes from the quick discussion between @puehringer and me. We will extend it later to make it more comprehensible.

* Set default column width in LineUp config (similar to `rowHeight` setting); at the moment it is fixed to 100px
* Extend `featureLevel` (`basic` and `advanced`) about custom feature levels
* Settings hierarchy: ranking, column type (e.g., string, number) or column instance (e.g., `string@current_club`), key (width) and of course the value
* Replace column annotations, e.g. `@toolbar(...)` or `@SortByDefault`, with getter/setter to allow extension/configuration at runtime

**Comment by sgratzl** [comment]
>     * Set default column width in LineUp config (similar to `rowHeight` setting); at the moment it is fixed to 100px

so far each column can determine its default column width. e.g. a string column has a default with of 200

>     * Replace column annotations, e.g. `@toolbar(...)` or `@SortByDefault`, with getter/setter to allow extension/configuration at runtime

what is the use case to configure one specific column instance vs a column type?

**Comment by thinkh** [comment]
General idea: at the moment we can configure column instances (e.g., via columnDesc), but we cannot configure the default columnDesc, which will be used by future instances

Example: at the moment a string column has a width of 200, but as a developer I want to set a default value for all columns (i.e., `Column` class) and override it for column types (e.g., `StringColumn`, `NumberColumn`) and even specific column instances (e.g., player, goals)

Our idea how to implement this, is to define a default column config, which will be inherited and can be extended. Maybe it is possible to integrate/use the existing columnDesc object?

// to be continued

**Comment by sgratzl** [comment]
> Example: at the moment a string column has a width of 200, but as a developer I want to set a default value for all columns (i.e., `Column` class) and override it for column types (e.g., `StringColumn`, `NumberColumn`) and even specific column instances (e.g., player, goals).

re width: in each column description for a column that one defines, one can specify the column width already. So, one can do the last case. The other cases could be implemented by merging different column descriptions when creating lineup. as in 

```js
col = { width: 100};
string = {width: 200}

age = {...col, column: 'age', width: 200}
player = {...col,...string, column: 'player', width: 300}
```

what is the gain on e.g. separating the with in an extra configuration object outside the descriptions?

**Comment by sgratzl** [comment]
Moreover, general question: are we talking about build time = initializing or run time = change while lineup is running?

**Comment by sgratzl** [comment]
> enable/disable all actions individually in toolbar (similar to current `featureLevel` implementation)

are we talking about top level menu icons, right? not sub level ones. What happens in cases in which e.g. one disables the grouping feature -> lineup has to be aware then also not showing it in group hierarchy widget. 

similar question what about bigger ui elements: sort hierarchy widget, group hierarchy, ...



> other interactions
> 
>     * is column movable

moveable what would it mean if one out of four columns is not moveable and e.g. the column is at the 2nd position?

**Comment by thinkh** [comment]
> what is the gain on e.g. separating the with in an extra configuration object outside the descriptions?

Our customers want to extend columns and create new columns. Hence, to us it is important that we can set the default configuration once and LineUp extends/passes the configuration internally.

> Moreover, general question: are we talking about build time = initializing or run time = change while lineup is running?

In our opinion, it is fine to set the default when initializing LineUp. We do not need to change the defaults at run time (because we can change each column individually).

> are we talking about top level menu icons, right? not sub level ones.

If we understand you correctly, then the top level menu icons are the shortcut icons and the sub level ones are in the more menu. 

![grafik](https://user-images.githubusercontent.com/5851088/76745112-995c0280-6775-11ea-8960-2100dae98fab.png)

Probably each feature needs the following configuration flags:

```
rankingIcon: true | 'shortcut' | 'more' | false
sidepanelIcon: true | 'shortcut' | 'more' | false
```

with

* `true`: Show icon in shortcut and more menu
* `shortcut`: Show icon only in shortcut menu
* `more`: Show icon only in more menu
* `false` Show no icon (= disable functionality)

The naming needs to be adjusted.
 
> What happens in cases in which e.g. one disables the grouping feature -> lineup has to be aware then also not showing it in group hierarchy widget.
> 
> similar question what about bigger ui elements: sort hierarchy widget, group hierarchy, ...

In this specific example a `groupable` property is missing on the columns. Currently it is [inferred from the toolbar annotation](https://github.com/lineupjs/lineupjs/blob/develop/src/ui/toolbar.ts#L422-L425), which is not programatically changable. And yes, the group hierarchy widget has to be aware if a column is `groupable` to create/update the dropdown.

> moveable what would it mean if one out of four columns is not moveable and e.g. the column is at the 2nd position?

If you have three columns `A | B | C` with `B` moveable and `A` and `C` are not. Then one can enforce that `A` is always in front of `C`, even though `B` could be moved before `A` or before and after `C`. I remember that we have the `frozen` and `fixed` properties, would they already solve the point? Otherwise could we deactivate the movement (drag'n'drop) of columns for the whole ranking so that it is static?

**Comment by sgratzl** [comment]
I've have been starting with a different approach for the column description customization, let me know what you think of it. 

the data provider has now a new centralized method which is responsible for creating column instances:

```ts
  protected instantiateColumn(type: IColumnConstructor, id: string, desc: IColumnDesc, typeFactory: ITypeFactory) {
    return new type(id, desc, typeFactory);
  }
```

Thus, when subclassing the LocalDataProvider one can override this method and apply defaults to the description or the column instance at ones will. 

e.g. in your example

```

    Column class has width: 100
    StringColumn has width: 200
    player@string has a width: 400
```
```ts
protected instantiateColumn(type: IColumnConstructor, id: string, desc: IColumnDesc, typeFactory: ITypeFactory) {
    if (desc.width != null) {
      // customized
    } else if (desc.type === 'string' && (<any>desc).column === 'player') {
       desc.width = 400;
    } else if (desc.type === 'string') {
       desc.width = 200;
    } else {
       desc.width = 100;
    }
    return new type(id, desc, typeFactory);
  }
```

**Comment by sgratzl** [comment]
regarding toolbar: 

instead of some magic assumptions, a column to list all toolbar icons. Tho, there are two exceptions: remove (if not fixed), and the change visualization type (since the renderer determine the menu and not the column). However, the determination whether a column can be sorted / grouped / group sorted is still based on the toolbar keys but in a more general way. 

there are two new options:  
```ts
 /**
   * hook for postprocess the toolbar actions for a column
   */
  resolveToolbarActions: (col: Column, keys: string[], lookup: IToolbarLookup) => IToolbarAction[];
  /**
   * hook for postprocess the toolbar dialog addons for a column
   */
  resolveToolbarDialogAddons: (col: Column, keys: string[], lookup: IToolbarLookup) => IToolbarDialogAddon[];
```
  which are responsible for resolving a list of toolbar keys of a column to a list of their action objects. One can customize it and apply some extra filtering, e.g. similar to the `canRender` function

regarding toolbar icons in sidepanel vs header
 the shortcut icon has a new option to provide a function which determines the shortcut state
```ts
shortcut: boolean | 'only' | ((col: Column, mode: 'sidePanel' | 'header') => boolean | 'only');
```
```
